python/pyspark/pandas/data_type_ops/datetime_ops.py

```diff
@@ -58,15 +66,18 @@ def sub(self, left: IndexOpsLike, right: Any) -> SeriesOrIndex:
             "The timestamp subtraction returns an integer in seconds, "
             "whereas pandas returns 'timedelta64[ns]'."
         )
-        if isinstance(right, IndexOpsMixin) and isinstance(right.spark.data_type, TimestampType):
+        if isinstance(right, IndexOpsMixin) and isinstance(
+            right.spark.data_type, (TimestampType, TimestampNTZType)
+        ):
             warnings.warn(msg, UserWarning)
             return left.astype("long") - right.astype("long")
```

Comment:
If left side is LTZ and right side is NTZ, we need to let Spark add implicit cast.

Reply:
This is to work around pandas' behaviour on internal types. PySpark didn't yet implement internal types so we can't let Spark do the implicit cast here.

Reply:
Not very familiar with pandas. Can you give a real example that will hit this code path?

Reply:
This is not related to the change. It will hit the code path when:

pd.Timestamp - pd.Timestamp which will result in interval types.

Once we implement ones in PySpark, we should switch it to return interval types. Currently, we return long values instead of interval types due to missing interval in PySpark.

Reply:
Yeah. So this one has to be removed once intervals are implemented in PySpark. At this moment, we cannot remove this or let Spark SQL to decide it by implicit cast. Not because Spark SQL does not have the implicit cast on NTZ and LTZ, but because PySpark doesn't have interval implementation.

Or do you suggest to implement the type coercion between NTZ and LTZ in this PR, and use something like `(right - left).astype("int")` in this PR?

Reply:
Do you suggest something like `(right - left).astype("int")`? This won't work because:
1. interval can't be converted to longs. To natively support this, it requires internal implementation on PySpark
2. `TimestampNTZ` is considered as unix timestamp in UTC but `TIMESTAMP_LZT - TIMESTAMP_NZT` or `TIMESTAMP_LZT - TIMESTAMP_NZT` will assume `TIMESTAMP_NZT` is in local session timezone. e.g.):
    ```scala
    scala> sql("SELECT TIMESTAMP '1970-01-01 00:00:00' - TIMESTAMP_NTZ '1970-01-01 00:00:00'").show(false)
    ```
    should result in something like `INTERVAL '0 09:00:00' DAY TO SECOND` (I am in KST) but it result in `INTERVAL '0 00:00:00' DAY TO SECOND`

Reply:
offline discussed.